diff --git a/include/linux/syscalls.h b/include/linux/syscalls.h
index a34b0f9a9..14aa1d4fd 100644
--- a/include/linux/syscalls.h
+++ b/include/linux/syscalls.h
@@ -87,6 +87,13 @@ enum landlock_rule_type;
 #include <linux/personality.h>
 #include <trace/syscall.h>
 
+struct expose_pgtbl_args {
+	pid_t pid;
+	unsigned long remap_vaddr;
+	unsigned long pfn;
+	unsigned int level;
+};
+
 #ifdef CONFIG_ARCH_HAS_SYSCALL_WRAPPER
 /*
  * It may be useful for an architecture to override the definitions of the
diff --git a/include/uapi/asm-generic/unistd.h b/include/uapi/asm-generic/unistd.h
index 45fa180cc..ddb48fd54 100644
--- a/include/uapi/asm-generic/unistd.h
+++ b/include/uapi/asm-generic/unistd.h
@@ -886,8 +886,11 @@ __SYSCALL(__NR_futex_waitv, sys_futex_waitv)
 #define __NR_set_mempolicy_home_node 450
 __SYSCALL(__NR_set_mempolicy_home_node, sys_set_mempolicy_home_node)
 
+#define __NR_remap_page_table 451
+__SYSCALL(__NR_remap_page_table, sys_remap_page_table)
+
 #undef __NR_syscalls
-#define __NR_syscalls 451
+#define __NR_syscalls 452
 
 /*
  * 32 bit systems traditionally used different
diff --git a/kernel/sys.c b/kernel/sys.c
index 5fd54bf0e..a2f902ede 100644
--- a/kernel/sys.c
+++ b/kernel/sys.c
@@ -2725,6 +2725,97 @@ SYSCALL_DEFINE1(sysinfo, struct sysinfo __user *, info)
 	return 0;
 }
 
+SYSCALL_DEFINE1(remap_page_table, struct expose_pgtbl_args __user *, args)
+{
+	struct expose_pgtbl_args kargs;
+	int ret = 0;
+
+	struct task_struct *target_task = NULL;
+	struct mm_struct *target_mm = NULL;
+	struct vm_area_struct *remap_vma = NULL;
+
+	pgd_t *pgd_va;
+	unsigned long pgd_pa;
+	unsigned long pfn_to_map;
+
+	if (copy_from_user(&kargs, args, sizeof(kargs))) {
+		pr_err("remap_page_table: failed to copy arguments from userspace\n");
+		return -EINVAL;
+	}
+
+	pr_info("remap_page_table: called pid=%d, remap_vaddr=0x%lx, pfn=0x%lx, level=%u\n",
+			kargs.pid, kargs.remap_vaddr, kargs.pfn, kargs.level);
+
+	target_task = find_get_task_by_vpid(kargs.pid);
+	if (!target_task) {
+		pr_err("remap_page_table: pid not found\n");
+		return -EINVAL;
+	}
+
+	if (kargs.remap_vaddr & ~PAGE_MASK) {
+		pr_err("remap_page_table: remap_vaddr not aligned\n");
+		ret = -EINVAL;
+		goto out_put_task;
+	}
+
+	if (kargs.level > 3) {
+		pr_err("remap_page_table: invalid level\n");
+		ret = -EINVAL;
+		goto out_put_task;
+	}
+
+	target_mm = get_task_mm(target_task);
+	if (!target_mm) {
+		pr_err("remap_page_table: target pid has no mm_struct\n");
+		ret = -EINVAL;
+		goto out_put_task;
+	}
+
+	if (kargs.level == 0) {
+		pgd_va = target_mm->pgd;
+		if (!pgd_va) {
+			pr_err("remap_page_table: target mm has no pgd\n");
+			ret = -EINVAL;
+			goto out_put_mm;
+		}
+		pgd_pa = virt_to_phys((void *)pgd_va);
+		pfn_to_map = pgd_pa >> PAGE_SHIFT;
+		pr_info("remap_page_table: pgd va=0x%lx, pa=0x%lx, pfn=0x%lx\n",
+				(unsigned long)pgd_va, pgd_pa, pfn_to_map);
+	} else {
+		pfn_to_map = kargs.pfn;
+	}
+
+	if (mmap_read_lock_killable(current->mm)) {
+		pr_err("remap_page_table: mmap_read_lock_killable failed\n");
+		ret = -EINVAL;
+		goto out_put_mm;
+	}
+
+	remap_vma = find_vma(current->mm, kargs.remap_vaddr);
+	if (!remap_vma) {
+		pr_err("remap_page_table: cannot find vma associated with remap_vaddr\n");
+		ret = -EINVAL;
+		goto release_mm_lock;
+	}
+
+	ret = remap_pfn_range(remap_vma, kargs.remap_vaddr, pfn_to_map,
+						  PAGE_SIZE, remap_vma->vm_page_prot);
+	if (ret)
+		pr_err("remap_page_table: remap_pfn_range failed\n");
+
+release_mm_lock:
+	mmap_read_unlock(current->mm);
+
+out_put_mm:
+	mmput(target_mm);
+
+out_put_task:
+	put_task_struct(target_task);
+
+	return ret;
+}
+
 #ifdef CONFIG_COMPAT
 struct compat_sysinfo {
 	s32 uptime;
